### PR TITLE
EMA Cross calculator and strategy ".py" file separate from main per tutorials

### PR DESCRIPTION
**Pull request by BensGitHub2022**
Updated file structure per project tutorial. Added EMA Cross calculator and separate module for the strategy in a new ".py" file.

Also slight modifications to mt5_lib.py to include console output for initialization and login - it will raise an exception if either initialization or login fails now. Let me know if this works for you guys.

No modifications to requirements.txt or README
**FYI:** I think we need to modify our README for the following:
- MetaTrader only works on Python up to 3.10, it will not work on 3.11 or 3.12, I found this out the hard way :-)
- I am pretty sure MetaTrader is only supported on Windows currently.

**Note:**
- I modified my settings.json as below (only one symbol):

```
{
    "mt5": {
        "login": 72467727,
        "password": "onkz0pjy",
        "server": "MetaQuotes-Demo",
        "terminal_pathway": "C:\\Program Files\\MetaTrader 5\\terminal64.exe",
        "symbols": ["USDSEK"],
        "timeout":60000,
        "timeframe": "M1"
    }
}
```

- I had to update ACCOUNT_SETTINGS_PATH!!!

From: ACCOUNT_SETTINGS_PATH = "../settings.json" did not work for me :(

To: ACCOUNT_SETTINGS_PATH = "settings.json"

### CODE REVIEW

**Comment by SaralynOgden** [comment]
image
Just FYI, there's a code option in the text menu that is great for formatting text as code. I updated your description to use it.

File: settings.json
Comment (SaralynOgden) [inline settings.json:1]: We want to make sure not to bring this in since it contains our login credentials. I see you added it to the .gitignore. You should also get it when you resolve the conflicts.

File: src/ema_cross_strategy.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    :param symbol: string of the sumbol to be queried
+    :param timeframe: string of the timeframe to be queried
+    :param ema_one: integer of the lowest timeframe length for EMA
+    :param ema_two: intefer of the highest timeframe length for EMA
+    """
+
+    ### Pseudo Code Steps
+    # Step 1: Retreive data -> get_data()
+    # Step 2: Calculated indicators -> calc_ind()
+    # Step 3: Determine if a trade event has occured -> det_trade()
```
Comment (SaralynOgden) [inline src/ema_cross_strategy.py:16]: We don't explicitly have this step in this process. Maybe reword this comment?

File: src/ema_cross_strategy.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    # Step 4: Return information back to the user
+
+    # Step 1
+    data = get_data(
```
Comment (SaralynOgden) [inline src/ema_cross_strategy.py:20]: Let's name this variable candles to make it clearer what type of data we are getting.

File: src/ema_cross_strategy.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+import indicator_lib
+import mt5_lib
+
+def ema_cross_strategy(symbol, timeframe, ema_one, ema_two):
```
Comment (SaralynOgden) [inline src/ema_cross_strategy.py:4]: Let's rename our ema_one and ema_two to reference the length of each. Something like short_term_ema and long_term_ema

File: src/ema_cross_strategy.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+def calc_ind(data, ema_one, ema_two):
+    """
+    Function to calculate the indicators for the EMA Cross strategy
+    :param data: dataframe of the raw data
+    :param ema_one: integer for the first ema
+    :param ema_two: integer for the second ema
+    :return: dataframe with updated columns
+    """
+    # Calculate the first EMA
+    dataframe = indicator_lib.calc_ema(
```
Comment (SaralynOgden) [inline src/ema_cross_strategy.py:42]: Let's rename these two dataframes to reference that they are for the short and long-term emas. Something like short_term_ema_dataframe and long_term_ema_dataframe

File: src/ema_cross_strategy.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        dataframe=data,
+        ema_size=ema_one
+    )
+    # Calculate the second ema
+    dataframe = indicator_lib.calc_ema(
+        dataframe=dataframe,
+        ema_size=ema_two
+    )
+    # Calculate the EMA Cross
+    dataframe = indicator_lib.ema_cross_calc(
```
Comment (SaralynOgden) [inline src/ema_cross_strategy.py:52]: Let's rename this to indicate that it's an ema cross dataframe. Something like ema_cross_dataframe

File: src/ema_cross_strategy.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+def get_data(symbol, timeframe):
+    """
+    Function to retreive data from MT5. Data is in the from of candlesticks and is retreived as a dataframe
+    :param symbol: string of the symbol to be retreived
+    :param timeframe: string of the timeframe to be retreived
+    :return: dataframe to user
+    """
+
+    # Retreive the data
+    data = mt5_lib.get_candlesticks(
```
Comment (SaralynOgden) [inline src/ema_cross_strategy.py:69]: Let's also name this candles here.

File: src/indicator_lib.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import numpy as np
+
+def calc_ema(dataframe, ema_size):
+    """
+    Calculates the Experimental Moving Average (EMA) of size `ema_size`
```
Comment (SaralynOgden) [inline src/indicator_lib.py]: *Exponential

File: src/indicator_lib.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
+import numpy as np
+
+def calc_ema(dataframe, ema_size):
```
Comment (SaralynOgden) [inline src/indicator_lib.py:7]: It looks like emas are always calculated based off of candles, so let's name dataframe here candles so as to clarify what the dataframe is.

File: src/mt5_lib.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+    #Get candles
+    candles = mt5.copy_rates_from_pos(symbol, mt5_timeframe, 1, num_candlesticks)
+
+    #return a pandas dataframe
+    return pandas.DataFrame(candles)
+
+
+
+def get_mt5_timeframe(timeframe):
```
Comment (SaralynOgden) [inline src/mt5_lib.py]: We can do this with enums and I think it'd be cleaner. https://docs.python.org/3/library/enum.html

File: src/ema_cross_strategy.py (hunk 1 above)
Comment (BensGitHub2022) [reply src/ema_cross_strategy.py:16]: To be added in the next coding stages! But yes this is totally delete-able or can be modified

**Comment by bpsix** [comment]
looks good, ill update the comments and variables after i merge

**Review by bpsix — APPROVED** [review]
lgtm
